In [1]:
import pandas as pd
import numpy as np

from scipy.sparse import csr_matrix
from scipy.sparse.linalg import svds

In [2]:
ratings = pd.read_csv('../ml-25m/ratings.csv')

print(ratings.shape)

(25000095, 4)


In [3]:
user_ids = ratings['userId'].unique()

user_map = {
    old:new
    for new, old in enumerate(user_ids)
}

movie_ids = ratings['movieId'].unique()

movie_map = {
    old:new
    for new, old in enumerate(movie_ids)
}

In [4]:
ratings['user_idx'] = ratings['userId'].map(user_map)

ratings['movie_idx'] = ratings['movieId'].map(movie_map)

In [5]:
R = csr_matrix(
    (
        ratings['rating'],
        (
            ratings['user_idx'],
            ratings['movie_idx']
        )
    )
)

print(R.shape)


(162541, 59047)


In [6]:
k = 50

In [7]:
U, sigma, Vt = svds(
    R,
    k=k
)

In [8]:
print("U Shape:", U.shape)
print("Sigma Shape:", sigma.shape)
print("Vt Shape:", Vt.shape)

U Shape: (162541, 50)
Sigma Shape: (50,)
Vt Shape: (50, 59047)


In [9]:
sigma

array([ 666.92696732,  670.73527139,  677.47019993,  683.81587672,
        685.85674261,  691.37311422,  698.78812156,  704.15116056,
        710.22849333,  722.47739601,  735.62205007,  743.16068133,
        750.43035299,  757.77330816,  765.09961141,  776.89898584,
        781.64514084,  795.88644892,  801.99440344,  817.61266693,
        831.75082441,  842.99038623,  855.12792653,  864.50608322,
        877.29464712,  892.38022653,  910.34900127,  934.29312008,
        966.2885601 ,  982.72741282, 1001.56918191, 1037.3438447 ,
       1063.47195891, 1082.55213676, 1116.4709135 , 1141.48115699,
       1234.31739957, 1273.31426397, 1296.83396612, 1342.67183332,
       1365.56008074, 1510.28047538, 1631.80146143, 1939.52362793,
       2004.32964045, 2194.9179759 , 2503.11071445, 2746.14854894,
       3673.71743681, 8037.03880765])

In [10]:
sigma_matrix = np.diag(sigma)

In [11]:
print(sigma_matrix.shape)

(50, 50)


In [12]:
print(sigma[:10])

[666.92696732 670.73527139 677.47019993 683.81587672 685.85674261
 691.37311422 698.78812156 704.15116056 710.22849333 722.47739601]


In [13]:
print("Largest Singular Value:", sigma.max())
print("Smallest Singular Value:", sigma.min())

Largest Singular Value: 8037.038807648901
Smallest Singular Value: 666.9269673189106


# Day 3 — Singular Value Decomposition (SVD)

## Objective
Apply Singular Value Decomposition (SVD) on the user-item matrix to discover latent user and movie factors.

---

## Tasks Completed

### 1. Sparse Matrix Preparation
Reconstructed the user-item matrix from MovieLens ratings data.

### 2. Selection of Latent Factors
Selected:

k = 50

latent dimensions for matrix factorization.

### 3. SVD Decomposition
Applied SciPy's svds() function.

The user-item matrix R was decomposed into:

R = UΣVᵀ

where:

- U = User latent factor matrix
- Σ = Singular values
- Vᵀ = Movie latent factor matrix

### 4. Matrix Shapes

Observed:

- U Shape = (162541, 50)
- Sigma Shape = (50,)
- Vᵀ Shape = (50, 59047)

### 5. Singular Value Analysis
Examined singular values to understand latent factor importance.

---

## Key Findings

1. Successfully factorized the sparse user-item matrix.
2. Reduced dimensionality from thousands of movie dimensions to 50 latent factors.
3. Captured hidden relationships between users and movies.
4. Prepared latent representations for recommendation generation.

---

## Deliverables Produced

- 03_SVD_Model.ipynb
- U Matrix
- Sigma Matrix
- Vᵀ Matrix

---

## Status

Day 3: COMPLETED ✅

Next Stage:
Day 4 — Top-N Recommendation Generation